### PR TITLE
[KED-1467] Added a new Dataset to fetch data from API endpoints

### PR DESCRIPTION
**Pull request by ppsimatikas**
## Description
Resolves: #283 

## Checklist

- [x] Read the [contributing](../CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [ ] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](../RELEASE.md) file
- [x] Added tests to cover my changes

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

**Comment by limdauto** [comment]
Hi @ppsimatikas thank you very much for this. This looks useful, but I'd like to know your opinion on whether some of the datasets that we provide out of the box:

* [JSONDataSet](https://kedro.readthedocs.io/en/latest/kedro.extras.datasets.pandas.JSONDataSet.html) also allows you to specify an HTTP URL as the filepath. You can encode the query parameters in the URL as well.
* If you require plain text from the API, [TextDataSet](https://kedro.readthedocs.io/en/latest/kedro.extras.datasets.text.TextDataSet.html) works in a similar way.

Some other considerations:

* In general, we prefer to use [fsspec](https://filesystem-spec.readthedocs.io/en/stable/) to interface with different data sources. I was wondering if their [HTTPFileSystem](https://filesystem-spec.readthedocs.io/en/stable/api.html#fsspec.implementations.http.HTTPFileSystem) could be of use here?
* IMHO, a generic API dataset seems too generic to be useful out of the box to many users without further modification. As an example, in my experience, most APIs are paginated and they are paginated differently. Some also follow a hypermedia format where you have to follow links indefinitely. So users will most certainly need to add further modification to this dataset if they want to use it. I'm just wondering whether it's worth adding a generic, base dataset in this case.

Regardless, this is a very valid use case, so another contribution idea if you are keen is to turn it into a tutorial on how to write a read-only API dataset. I think we would be happy to include it in our docs along side the [Creating a new dataset tutoria](https://kedro.readthedocs.io/en/latest/04_user_guide/14_create_a_new_dataset.html). What do you think?

**Comment by ppsimatikas** [comment]
@limdauto,

Thank you for your quick reply!

As a data engineer I am many times tasked of getting data from external data providers via an API.
I feel like a generic dataset such the one I have in this PR will be super useful for my current and future engagements.

I have looked at both JSON and Text DataSets. They only access a URL via a GET request. You have no control over the request attributes, method, headers, auth, etc...
Most APIs out there, they at least require some kind of authentication headers.

I took a look at the HTTPFileSystem from fsspec. It seems like it has the exact limitations I mentioned above, similar to JSON / Text DataSets. https://filesystem-spec.readthedocs.io/en/stable/_modules/fsspec/implementations/http.html#HTTPFileSystem

I agree, many APIs have totally different structures and paginations. That is the case for many of the data sources in Kedro. For example the pandas.SQLQueryDataSet, you usually never just blindly run a static SQL query. You will most likely require require to add some dynamic filtering, for example, start and end dates.

I have solved that requirement by creating my own datasource while extending the SQLQueryDataSet. Example:

`
class MySQLDataSet(SQLQueryDataSet):
    def __init__(
            self, credentials: Dict[str, Any], load_args: Dict[str, Any] = None
    ) -> None:
        super().__init__(self._get_sql(), credentials, "ORACLE", load_args)

    def _get_sql(self) -> str:
        start_date = datetime.now() - 1 month
        end_date = datetime.now()
        return "select * from my_table where date is between " + start_date + " and " + end_date
`

With similar mind set for more complicated API calls the developer can extend my APIDataSet and as creative as the desire !

File: tests/extras/datasets/api/test_api.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+TEST_PARAMS = {"param": "value"}
+TEST_URL_WITH_PARAMS = TEST_URL + "?param=value"
+
+TEST_HEADERS = {"key": "value"}
+
+
+@requests_mock.Mocker()
+class TestAPIDataSet(unittest.TestCase):
+    def test_successfully_load_with_text_response(self, requests_mocker):
+        for method in POSSIBLE_METHODS:
```
Comment (DmitriiDeriabinQB) [inline tests/extras/datasets/api/test_api.py]: Can you please use `pytest.mark.parametrize` as in [test_biosequence_dataset](https://github.com/quantumblacklabs/kedro/blob/9809bd7ca0556531fa4a2fc02d5b2dc26cf8fa97/tests/extras/datasets/bioinformatics/test_biosequence_dataset.py#L94), for example?

**Comment by DmitriiDeriabinQB** [comment]
@ppsimatikas, please let me know if you would like to go ahead and address the comments above. Otherwise, I'm happy to take over and finish it up.

**Comment by ppsimatikas** [comment]
> @ppsimatikas, please let me know if you would like to go ahead and address the comments above. Otherwise, I'm happy to take over and finish it up.

I am about to take care of them :) Thank you for ensuring this goes through !

File: tests/extras/datasets/api/test_api.py (hunk 1 above)
Comment (ppsimatikas) [reply tests/extras/datasets/api/test_api.py:55]: This is a neat trick !

**Comment by ppsimatikas** [comment]
@DmitriiDeriabinQB All comments have been addressed.

File: kedro/extras/datasets/api/api.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        )
+
+    def _exists(self) -> bool:
+        try:
+            response = self._execute_request()
+        except DataSetNotFoundError:
+            return False
+
+        # NOTE: we don't access the actual content here, which might be large.
+        return response.status_code == requests.codes.OK  # pylint: disable=no-member
```
[the remark this answers is not in the document]
Comment (DmitriiDeriabinQB) [reply kedro/extras/datasets/api/api.py]: I think this is still relevant

[the remark this answers is not in the document]
Comment (ppsimatikas) [reply kedro/extras/datasets/api/api.py]: This is fixed.

[the remark this answers is not in the document]
Comment (DmitriiDeriabinQB) [reply kedro/extras/datasets/api/api.py]: Perfect, thank you!

**Comment by 96imranahmed** [comment]
Hey team 😄 ! Sorry for reviving a dead thread, but have we considered extending this dataset to support API 'save' operations: e.g., posting pipeline outputs to another endpoint? I have several systems (e.g., pricing, recommenders) that need to push API requests to various other microservices in our stack. 

I was wondering whether I should simply extend the API Dataset, or whether there was another best-practice way of handling these 'live' pipeline outputs?

**Comment by limdauto** [comment]
@96imranahmed Hey, how are you doing? Are you using Kedro at Pathao? 

I think what you are looking for the `after_pipeline_run` hook: https://kedro.readthedocs.io/en/0.17.0/kedro.framework.hooks.specs.PipelineSpecs.html#kedro.framework.hooks.specs.PipelineSpecs.after_pipeline_run -- you get access to the `run_result` in that hook, so you can just instantiate an HTTP request with that payload. If you want to do it on a per-node basis, you can use the `after_node_run` hook. It fits better than piggybacking on the APIDataSet

**Comment by 96imranahmed** [comment]
Hey @limdauto - doing well! Yes, trying to get the team to push to using Kedro for data-pipelining of our ML model stack (as opposed to doing everything through SQL). 

Fabulous - I didn't know these existed! Will get this hooked-up now :) Thanks!

**Pull request merged**